Adress zero index tl issue

SharkyZg opened the pull request:
When the car reaches end of the track in the simulator waypoint index resets to 0. That would cause problems for traffic light detection before this fix because traffic lights are detected based on difference between waypoint index of car and waypoint index of traffic light.

I have addressed this issue and exported closest traffic light detection code to another get_closest_tl_wp_idx function to make the code more readable.

SharkyZg committed:
Adressing waypoint zero index traffic light issue

When the car reaches end of the track in the simulator
waypoint index resets to 0. That would cause problems for
traffic light detection before this fix because
traffic lights are detected based on the difference
between waypoint index of car nad waypoint index of traffic
light.

SharkyZg committed:
Adressing waypoint zero index traffic light issue

When the car reaches end of the track in the simulator
waypoint index resets to 0. That would cause problems
for traffic light detection before this fix because
traffic lights are detected based on the difference
between waypoint index of car nad waypoint index of
traffic light.

I have addressed this issue and exported closest
traffic light detection code to another get_closest_tl_wp_idx
function to make the code more readable.

SharkyZg commented:
Please squash and merge commits if you decide to merge.

SharkyZg committed:
Merge remote-tracking branch 'upstream/master' into adress_zero_index_tl_issue

carlosgalvezp commented:
Thanks! I'll have a look tomorrow :)

idanov commented on ros/src/tl_detector/tl_detector.py:
```diff
@@ -269,18 +269,63 @@ def get_light_state(self, light):
+            # Cover corner case where traffic light is just behind the
+            # waypoint 0 And the car is near last waypoint so waypoint
+            # index will reset to 0
+            if(searching_dist_tl > (car_distance_from_last_wp) and
```
Here the 80 characters line limit makes the code hard to read... Could we bump up the limit to 120 characters? @carlosgalvezp

idanov commented on ros/src/tl_detector/tl_detector.py:
```diff
@@ -293,33 +338,25 @@ def process_traffic_lights(self):
+            light = self.waypoints.waypoints[light_wp_idx]
             state = self.get_light_state(light)
-            return light_wp, state
+            print(light_wp_idx)
```
Could we `rospy.loginfo()` here?

SharkyZg replied:
@idanov sure, or I can remove this line completely, I've forgot to remove it after testing.

idanov reviewed (approved):
Great, LGTM! I didn't checkout and test it, but I assume it works as expected. I will test it tomorrow if you haven't merged it yet.

SharkyZg committed:
Adressing waypoint zero index traffic light issue

When the car reaches end of the track in the simulator
waypoint index resets to 0. That would cause problems
for traffic light detection before this fix because
traffic lights are detected based on the difference
between waypoint index of car nad waypoint index of
traffic light.

I have addressed this issue and exported closest
traffic light detection code to another get_closest_tl_wp_idx
function to make the code more readable.

carlosgalvezp commented on ros/src/tl_detector/tl_detector.py:
```diff
@@ -268,18 +268,76 @@ def get_light_state(self, light):
+        # create it
+        if(self.tl_waypoints_idx == []):
+            self.get_tl_waypoints_idx()
+        last_waypoint_idx = len(self.waypoints.waypoints) - 1
```
When passed to the function `get_distance_in_track`, you should pass the actual length (as number of waypoints), not the index to the last element. So it should not have the `-1` there. Maybe you can call it `track_length` or similar as well?

Also, `car_distance_from_last_wp` is not used anymore so can be removed

carlosgalvezp reviewed (commented):
Looks really good! Just a minor comment

SharkyZg committed:
Adressing waypoint zero index traffic light issue

When the car reaches end of the track in the simulator
waypoint index resets to 0. That would cause problems
for traffic light detection before this fix because
traffic lights are detected based on the difference
between waypoint index of car nad waypoint index of
traffic light.

I have addressed this issue and exported closest
traffic light detection code to another get_closest_tl_wp_idx
function to make the code more readable.
